smart-server/src/main/java/org/smartdata/server/actions/ActionRegister.java

```diff
@@ -0,0 +1,125 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.smartdata.server.actions;
+
+import org.apache.hadoop.conf.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.smartdata.common.actions.ActionType;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Provide Action Map and Action instance
+ */
+public class ActionRegister {
+  static final Logger LOG = LoggerFactory.getLogger(ActionRegister.class);
+
+  private static ActionRegister instance = new ActionRegister();
+  HashMap<String, Class> actionMap;
+
+  synchronized public static ActionRegister getInstance() {
+    return instance;
+  }
+
+  private ActionRegister() {
+    actionMap = new HashMap<>();
+  }
+
+  public void initial(Configuration conf) {
+    loadNativeAction();
+    try {
+      // TODO read path and map from configure
+      // String libPath = getClass().getClassLoader().getResource("UserAction").getPath();
+      loadUserDefinedAction(null, null);
+    } catch(Exception e) {
+      LOG.error(e.getMessage());
+      LOG.error("Load User Actions from Path error!");
+    }
+  }
+
+  public void clear() {
+    actionMap.clear();
+  }
+
+  public void loadNativeAction() {
+    // Note that native actions and ActionRegister are in the same path
+    String path = this.getClass().getCanonicalName();
+    path = path.substring(0, path.lastIndexOf('.') + 1);
+    LOG.info("Current Class Package {}", path);
+    for (ActionType t : ActionType.values()) {
+      String name = t.toString();
+      try {
+        actionMap.put(name, Class.forName(path + name));
+      } catch (ClassNotFoundException e) {
+        LOG.info("Class {} not found!", name);
+        continue;
+      }
+    }
+  }
+
+  public void loadUserDefinedAction(String libPath, Map<String, String> userDefinedActions) throws IOException {
+    File dependencyDirectory = new File(libPath);
+    File[] files = dependencyDirectory.listFiles();
+    if (files == null) {
+      return;
+    }
+    // Search this dir
+    for (File jarFile: files) {
+      if(jarFile.getName().contains("jar")) {
+        URLClassLoader classLoader = new URLClassLoader(new URL[]{new URL("file:" +
```

Comment:
I think all the jar files under the same directory should be included in a same classcloader